### PR TITLE
Allow JSON-representation of PropertyManager loaded

### LINKED ISSUE #23988: Correctly Deserialise PropertyManager Property
When project recovery saves out algorithm calls it currently serializes the input values into strings. If an algorithm takes a PropertyManager property however it is not currently deserialized properly leading to errors. 

I think this is occurring because when when the property manager property is created from the serialised string it tries to assign the properties values before declaring them. 

This can be seen in this method 

https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/PropertyManagerProperty.cpp#L74

which creates a new PropertyManager object and then tries to set Properties on it before declaring them.

So as not to replicate the json parsing code which already exists in PropertyManager.cpp it was suggested that this method

https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/PropertyManager.cpp#L223

be given an optional argument/provide an overide such that properties could be declared and set in one go. This would probably involve putting this new flag in several of the setProperties methods in this class and then at the point where properties are set if the flag is true also declare them first.

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Adds methods to decode/encode property values as JSON objects. Previously the encoding converted call of the values to strings whereas now the type is preserved. Decoding has also been added so that a `PropertyManager` can be reconstructed from its JSON representation. This is important for algorithms such as SetSample that accept `PropertyManager` objects as input and also for many of the SANS workflow algorithms.

**To test:**

The best way to test this is use the `SetSample` algorithm in conjuction with `MonteCarloAbsorption`.

Run the following

```python
data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target="Wavelength")
# Default up axis is Y
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07})
# Simulating every data point can be slow. Use a smaller set and interpolate
abscor = MonteCarloAbsorption(data, NumberOfWavelengthPoints=50)
```

and generate the history of `abscor`. Rename `abscor` to `abscor_orig` and then run the *generated* script from the history to produce a new `abscor`. Run `CompareWorkspaces` and they should match.

Fixes #23988 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/test/CMakeLists.txt [hunk 1]
```diff
@@ -23,7 +23,7 @@ if ( CXXTEST_FOUND )
 )
   add_dependencies ( FrameworkTests KernelTest )
   # Test data
-  add_dependencies ( KernelTest StandardTestData )
+  #add_dependencies ( KernelTest StandardTestData )
```
Comment (Mantid-Matthew) [inline Framework/Kernel/test/CMakeLists.txt]: Is this supposed to be commented out?

Comment (martyngigg) [reply Framework/Kernel/test/CMakeLists.txt]: Nice catch! That absolutely should not be commented out.

**Comment by Mantid-Matthew** [comment]
The given example works for me.

I find that the SANS workflow still does not work though. This appears to be because the SANSState object which many of the SANS algorithms take is a PropertyManagerProperty object which contains other PropertyManagerProperty objects as it is used to pass in a nested python dict. When this is serialised and deserialised on the top layer PM object seems to be being converted back into a PropertyManagerProperty object but lower layers seem to be left as a string.

I'm uncertain whether fixing this is within the scope of this PR though as it may be better to get these changes in which work for single layer objects and add an issue to fix nested PM objects later?

**Comment by martyngigg** [comment]
@Mantid-Matthew Thanks for testing this. I think I know why this is and I think it's worth fixing here. I will look into it

**Comment by martyngigg** [comment]
I think that issue with the nested property managers should now be fixed. I've also added a unit test for that.

**Comment by Mantid-Matthew** [comment]
I am currently getting python errors when I attempt to launch mantid with the latest version of this. 

```
Welcome to Mantid 3.13.20190126.953
Please cite: http://dx.doi.org/10.1016/j.nima.2014.07.029 and this release: http://dx.doi.org/10.5286/Software/Mantid
Could not open library /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: undefined symbol: _ZNK6Mantid6Kernel17PropertyWithValueIN5boost10shared_ptrINS0_15PropertyManagerEEEE5cloneEv
Could not open library /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: undefined symbol: _ZNK6Mantid6Kernel17PropertyWithValueIN5boost10shared_ptrINS0_15PropertyManagerEEEE5cloneEv
Could not open library /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: /home/matthew/Documents/Mantid/mantidBuild/bin/libMantidDataHandling.so: undefined symbol: _ZNK6Mantid6Kernel17PropertyWithValueIN5boost10shared_ptrINS0_15PropertyManagerEEEE5cloneEv
Traceback (most recent call last):
  File "<string>", line 29, in <module>
  File "/home/matthew/Documents/Mantid/mantidBuild/bin/mantidplot.py", line 17, in <module>
    import pymantidplot
  File "/home/matthew/Documents/Mantid/mantidBuild/bin/pymantidplot/__init__.py", line 24, in <module>
    import pymantidplot.proxies as proxies
  File "/home/matthew/Documents/Mantid/mantidBuild/bin/pymantidplot/proxies.py", line 22, in <module>
    import mantid
  File "/home/matthew/Documents/Mantid/mantidBuild/bin/mantid/__init__.py", line 97, in <module>
    from . import kernel as _kernel
  File "/home/matthew/Documents/Mantid/mantidBuild/bin/mantid/kernel/__init__.py", line 54, in <module>
    from ._kernel import *
ImportError: /home/matthew/Documents/Mantid/mantidBuild/bin/mantid/kernel/_kernel.so: undefined symbol: _ZNK6Mantid6Kernel17PropertyWithValueIN5boost10shared_ptrINS0_15PropertyManagerEEEE5cloneEv
Error running init file "/home/matthew/Documents/Mantid/mantidBuild/bin/mantidplotrc.py"
Previous initialization of Python failed, cannot retry.
```

**Comment by martyngigg** [comment]
@Mantid-Matthew This should finally be ready for testing again.

**Review by peterfpeterson — APPROVED** [review]
Massive changeset due to adding new functionality into `IPropertyManager`. This will make using algorithms from python significantly easier.

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could someone merge this on the command line to avoid a whole rebuild just for a release note fix?

**Pull request merged**